pkg/trace/metrics/timing/timing.go

```diff
@@ -0,0 +1,154 @@
+// Package timing is used to aggregate timing calls within hotpaths to avoid using
+// repeated statsd calls. The package has a default set that reports at 10 second
+// intervals and can be used directly. If a different behaviour or reporting pattern
+// is desired, a custom Set may be created.
+package timing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/atomic"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+)
+
+// AutoreportInterval specifies the interval at which the default set reports.
+const AutoreportInterval = 10 * time.Second
+
+var (
+	defaultSet = NewSet()
+	stopReport = defaultSet.Autoreport(AutoreportInterval)
+)
+
+// Since records the duration for the given metric name as time passed since start.
+// It uses the default set which is reported at 10 second intervals.
+func Since(name string, start time.Time) { defaultSet.Since(name, start) }
+
+// Stop permanently stops the default set from auto-reporting and flushes any remaining
+// metrics. It can be useful to call when the program exits to ensure everything is
+// submitted.
+func Stop() { stopReport() }
+
+// Set represents a set of metrics that can be used for timing. Use NewSet
+// to create one and Report (or Autoreport) to submit metrics. It is safe for
+// concurrent use.
+type Set struct {
+	mu sync.RWMutex        // guards c
+	c  map[string]*counter // maps names to their aggregates
+}
+
+// NewSet returns a new Set, optionally initialized with the given metric names.
+func NewSet(names ...string) *Set {
+	set := Set{
+		c: make(map[string]*counter, len(names)),
+	}
+	for _, name := range names {
+		set.c[name] = newCounter(name)
+	}
+	return &set
+}
+
+// Autoreport enables autoreporting of the Set at the given interval. It returns a
+// cancellation function.
+func (s *Set) Autoreport(interval time.Duration) (cancelFunc func()) {
+	stop := make(chan struct{})
+	go func() {
+		defer close(stop)
+		tick := time.NewTicker(interval)
+		defer tick.Stop()
+		for {
+			select {
+			case <-tick.C:
+				s.Report()
+			case <-stop:
+				s.Report()
+				return
+			}
+		}
+	}()
+	var once sync.Once // avoid panics
+	return func() {
+		once.Do(func() {
+			stop <- struct{}{}
+			<-stop
+		})
+	}
+}
+
+// Since records the duration for the given metric name as time passed since start.
+// If name does not exist, as defined by NewSet, it creates it.
+func (s *Set) Since(name string, start time.Time) {
+	s.mu.RLock()
```

Comment:
Minor suggestion: could put this "lock, read, lock" in `getOrCreateCounter` as well, then the full locking logic is contained within `getOrCreateCounter` and here you just call that to get a counter correctly.